src/main/java/network/brightspots/rcv/ClearBallotCvrReader.java

```diff
@@ -0,0 +1,144 @@
+ * the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License along with this
+ * program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package network.brightspots.rcv;
+
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import javafx.util.Pair;
+
+import network.brightspots.rcv.CastVoteRecord.CvrParseException;
+
+
+class ClearBallotCvrReader {
+
+  private final String cvrPath;
+  private final ContestConfig contestConfig;
+
+  ClearBallotCvrReader(String cvrPath, ContestConfig contestConfig) {
+    this.cvrPath = cvrPath;
+    this.contestConfig = contestConfig;
+  }
+
+  // parse Cvr json into CastVoteRecord objects and append them to the input castVoteRecords list
+  // see Clear Ballot 2.1 RCV Format Specification for details
+void readCastVoteRecords(List<CastVoteRecord> castVoteRecords) throws CvrParseException {
+    BufferedReader csvReader;
+    try {
+      csvReader = new BufferedReader(new FileReader(this.cvrPath));
+      // each "choice column" in the input Csv corresponds to a particular ranking: candidate+rank pair
+      // we parse these rankings from the header row into a map for lookup during CVR parsing
+      String firstRow = csvReader.readLine();
+      if (firstRow == null) {
+        Logger.log(Level.SEVERE, "No header row found in cast vote record file: %s", this.cvrPath);
+        throw new CvrParseException();
+      }
+      String[] headerData = firstRow.split(",");
+      if (headerData.length < CvrColumnField.ChoicesBegin.ordinal()) {
+        Logger.log(Level.SEVERE, "No choice columns found in cast vote record file: %s", this.cvrPath);
+      }
+      Map<Integer, Pair<Integer, String>> columnIndexToRanking = new HashMap<>();
+      for (int columnIndex = CvrColumnField.ChoicesBegin.ordinal(); columnIndex < headerData.length; columnIndex++) {
+        String choiceColumnHeader = headerData[columnIndex];
+        String[] choiceFields = choiceColumnHeader.split(":");
+        // filter by contest
+        String contestName = choiceFields[RcvChoiceHeaderField.CONTEST_NAME.ordinal()];
+        if (!contestName.equals(this.contestConfig.getContestId())) {
+          continue;
+        }
+        // validate and store the ranking associated with this choice column
+        String choiceName = choiceFields[RcvChoiceHeaderField.CHOICE_NAME.ordinal()];
+        if (!this.contestConfig.getCandidateCodeList().contains(choiceName)) {
+          Logger.log(Level.SEVERE, "Candidate: %s from cast vote record not found in config!", choiceName);
+          throw new CvrParseException();
+        }
+        Integer rank = Integer.parseInt(choiceFields[RcvChoiceHeaderField.RANK.ordinal()]);
```

Comment:
we should validate that this doesn't exceed the maxRankingsAllowed value set in the config

Reply:
Yeah good idea.